middle_end/flambda/types/env/aliases.ml

```diff
@@ -287,200 +287,325 @@ end = struct
+      match Name.Set.min_elt_opt symbols with
+      | Some symbol ->
+        Some (Simple.name symbol)
+      | None ->
+        match Name.Set.min_elt_opt vars with
+        | Some var ->
+          Some (Simple.name var)
+        | None ->
+          None
+end
+
+
 type t = {
-  canonical_elements : Simple.t Simple.Map.t;
+  canonical_elements : Simple.t Name.Map.t;
   (* Canonical elements that have no known aliases are not included in
      [canonical_elements]. *)
-  aliases_of_canonical_elements : Aliases_of_canonical_element.t Simple.Map.t;
-  (* For [elt |-> aliases] in [aliases_of_canonical_elements], then
+  aliases_of_canonical_names : Aliases_of_canonical_element.t Name.Map.t;
+  (* For [elt |-> aliases] in [aliases_of_canonical_names], then
      [aliases] never includes [elt]. *)
   (* CR mshinwell: check this always holds *)
-  binding_times_and_modes : Binding_time.With_name_mode.t Simple.Map.t;
+  aliases_of_consts : Aliases_of_canonical_element.t Const.Map.t;
+  binding_times_and_modes : Binding_time.With_name_mode.t Name.Map.t;
   (* Binding times and name modes define an order on the elements.
      The canonical element for a set of aliases is always the minimal
      element for this order, which is different from the order used
      for creating sets and maps. *)
 }
 
-let print ppf { canonical_elements; aliases_of_canonical_elements;
-                binding_times_and_modes; } =
+let print ppf { canonical_elements; aliases_of_canonical_names;
+                aliases_of_consts; binding_times_and_modes; } =
   Format.fprintf ppf
     "@[<hov 1>(\
       @[<hov 1>(canonical_elements@ %a)@]@ \
-      @[<hov 1>(aliases_of_canonical_elements@ %a)@]@ \
+      @[<hov 1>(aliases_of_canonical_names@ %a)@]@ \
+      @[<hov 1>(aliases_of_consts@ %a)@]@ \
       @[<hov 1>(binding_times_and_modes@ %a)@]\
       )@]"
-    (Simple.Map.print Simple.print) canonical_elements
-    (Simple.Map.print Aliases_of_canonical_element.print)
-    aliases_of_canonical_elements
-    (Simple.Map.print Binding_time.With_name_mode.print)
+    (Name.Map.print Simple.print) canonical_elements
+    (Name.Map.print Aliases_of_canonical_element.print)
+    aliases_of_canonical_names
+    (Const.Map.print Aliases_of_canonical_element.print)
+    aliases_of_consts
+    (Name.Map.print Binding_time.With_name_mode.print)
     binding_times_and_modes
 
-let defined_earlier t alias ~than =
-  let info1 = Simple.Map.find alias t.binding_times_and_modes in
-  let info2 = Simple.Map.find than t.binding_times_and_modes in
+let name_defined_earlier t alias ~than =
+  let info1 = Name.Map.find alias t.binding_times_and_modes in
+  let info2 = Name.Map.find than t.binding_times_and_modes in
   Binding_time.strictly_earlier
     (Binding_time.With_name_mode.binding_time info1)
     ~than:(Binding_time.With_name_mode.binding_time info2)
 
+let defined_earlier t alias ~than =
+  Simple.pattern_match than
```

Comment:
Yes, all aliases are considered to be defined at the same time, namely before anything else. (It doesn't really come up, since of course two constants that are aliases must actually be equal.)